pkg/codegen/jenny_crd_yaml.go

```diff
@@ -0,0 +1,213 @@
+// TypeMeta and ObjectMeta, as those do not contain YAML tags and get improperly serialized to YAML.
+// Since we don't need to use it with the kubernetes go-client, we don't need the extra functionality attached.
+//
+//nolint:lll
+type customResourceDefinition struct {
+	Kind       string                              `json:"kind,omitempty" yaml:"kind,omitempty" protobuf:"bytes,1,opt,name=kind"`
+	APIVersion string                              `json:"apiVersion,omitempty" yaml:"apiVersion,omitempty" protobuf:"bytes,2,opt,name=apiVersion"`
+	Metadata   customResourceDefinitionMetadata    `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+	Spec       k8ssys.CustomResourceDefinitionSpec `json:"spec"`
+}
+
+type customResourceDefinitionMetadata struct {
+	Name string `json:"name,omitempty" yaml:"name" protobuf:"bytes,1,opt,name=name"`
+	// TODO: other fields as necessary for codegen
+}
+
+type cueOpenAPIEncoded struct {
+	Components cueOpenAPIEncodedComponents `json:"components"`
+}
+
+type cueOpenAPIEncodedComponents struct {
+	Schemas map[string]any `json:"schemas"`
+}
+
+func valueToCRDSpecVersion(str string, name string, stored bool) (k8ssys.CustomResourceDefinitionSpecVersion, error) {
+	// Decode the bytes back into an object where we can trim the openAPI clutter out
+	// and grab just the schema as a map[string]any (which is what k8s wants)
+	back := cueOpenAPIEncoded{}
+	err := goyaml.Unmarshal([]byte(str), &back)
+	if err != nil {
+		return k8ssys.CustomResourceDefinitionSpecVersion{}, err
+	}
+	if len(back.Components.Schemas) != 1 {
+		// There should only be one schema here...
+		// TODO: this may change with subresources--but subresources should have defined names
+		return k8ssys.CustomResourceDefinitionSpecVersion{}, fmt.Errorf("version %s has multiple schemas", name)
+	}
+	var def map[string]any
+	for _, v := range back.Components.Schemas {
+		ok := false
+		def, ok = v.(map[string]any)
+		if !ok {
+			return k8ssys.CustomResourceDefinitionSpecVersion{},
+				fmt.Errorf("error generating openapi schema - generated schema has invalid type")
+		}
+	}
+
+	return k8ssys.CustomResourceDefinitionSpecVersion{
+		Name:    name,
+		Served:  true,
+		Storage: stored,
+		Schema: map[string]any{
+			"openAPIV3Schema": map[string]any{
+				"properties": map[string]any{
+					"spec": def,
+				},
+				"required": []any{
+					"spec",
+				},
+				"type": "object",
+			},
+		},
+	}, nil
+}
+
+func versionString(version thema.SyntacticVersion) string {
+	return fmt.Sprintf("v%d-%d", version[0], version[1])
+}
+
+// Hoisting this out of thema until we resolve the proper approach there
+func generateOpenAPI(sch thema.Schema, props kindsys.CoreStructuredProperties) (string, error) {
+	ctx := sch.Underlying().Context()
+	v := ctx.CompileString(fmt.Sprintf("#%s: _", props.Name))
+	defpath := cue.MakePath(cue.Def(props.Name))
+	defsch := v.FillPath(defpath, sch.Underlying())
+
+	cfg := &openapi.Config{
```

Comment:
In order to fix the issue with CRD generation that you're getting in https://github.com/grafana/grafana/pull/60207/files#r1046954435 you should pass `ExpandReferences: true` in the config (see https://pkg.go.dev/cuelang.org/go/encoding/openapi#Config for more details).

That way the generator would produce the structural OpenAPI schema subset that Kubernetes expects to see in the CRD schema definition.

Reply:
Unfortunately, that doesn't work - it was the first thing i tried in condensing all the types into a single schema. 

```
go generate ./kinds/gen.go
core kinddirs codegen failed: 1 error occurred:
	* YamlCRDJenny: unsupported op for number & for input "dashboard"
```

The `NameFunc` that's in there _should_ have the same effect...? i haven't investigated deeply. I'm also adapting this approach into thema itself - it's necessary to handle the `import` cases that we have in `models.cue` files [like these](https://github.com/grafana/grafana/blob/778337e522590f0758450e9948365e2958caafbc/public/app/plugins/panel/barchart/models.cue#L19).

In any case, the [complaints](https://gist.github.com/sdboyer/c52ffbb948978809c3bd13ac14b8a464) the apiserver fired back from the gist _appeared_ to have more to do with internal aspects of the schema...i think? idk, it was pretty overwhelming :P

Reply:
IIUC the complaints are exactly the same ones that we had to deal with, hence my comment. codegen error looks like our own issue?

Reply:
Codegen error probably is an internal thing with Thema, yes - it's one i've seen before, but only ephemerally while iterating on impl

i reformatted the error to be more legible. It seems to really not like disjunctions, which i believe was the thing that caused a problem before with SLOs

Reply:
One-ofs are definitely not supported well in CRDs, but they should still be translatable – we've solved that for SLO using `ExpandReferences: true`, so I'm not 100% sure why it's failing here.

Reply:
nor i - which is why i figured it's valuable to have this fallback approach of letting the kind spec control whether what amounts to a passthrough CRD should be generated